### ISSUE TITLE
nextjs example page3 serving wrong language during ssr

### ISSUE THREAD

**Issue by Enalmada** [i18next/react-i18next#574]
I am trying to get started with nextjs and react-i18next and hitting a problem that has me completely blocked.  I would like to pass "t" down from my _app.js to pages without wrapping each individual page in "export default withI18next()".    

essentially in _app.js
```
<NamespacesConsumer {...pageProps} ns="common" i18n={(pageProps && pageProps.i18n) || i18n} wait={process.browser} >
          {t => ( <Component t={t} {...pageProps} /> )}
</NamespacesConsumer>
```

Then pages are pure components:
```
export default const Page2 = ({t}) => (
        <h1>{t("welcomePage2")}</h1>
);
```

To reproduce:
Start react-i18next/example/nextjs
switch language to de
navigate to page 3: http://localhost:3000/de/page3
do a browser refresh of the page and note content is in en first and flickers to de
view source - the h1 from _app.js wrapped with "NamespacesConsumer" is in en not de

This is essentially my bug...the `<h1>{t('common:integrates_react-i18next')}</h1>` in _app.js seems correctly wrapped in NamespacesConsumer and should be rendering proper locale regardless of what pages under are doing.

Could you please help fix the example so that the page3 example renders the _app.js content in the correct language while keeping it a pure component?

**Comment by jamuhl** [comment]
did you move all translations to common.json?

as you only load that in the _app.js

`<NamespacesConsumer {...pageProps} ns="common" `

**Comment by jamuhl** [comment]
not even sure this is the way it works i guess the flow with app.js -> i guess that gets wrapped magically were in reality the page is the outer component (as pageProps come from the page)...

but honestly i'm far from an expert in next.js - so most parts are magic to me

**Comment by Enalmada** [comment]
@jamuhl I really appreciate you taking the time to help me out despite your limited experience with it.    In my project I am ok with everything being in common btw as long as ssr works in pure page components.  Fortunately this issue is very easy to reproduce...fire up the nextjs example, click button to change to de, click page3 link, do a browser refresh on page and you will see the flicker and page source being in english.   I feel like putting the `export default withI18next(['common'])(MyApp)` in _app should work and fix this...I assume there is some magic that withI18next is doing to the pages that NamespacesConsumer isn't...otherwise the t("") call from _app.js would look fine on page3 refresh.  Perhaps the circular dependency cause is worth looking into?

**Comment by jamuhl** [comment]
oh...yes...withI18next -> does https://github.com/i18next/react-i18next/blob/master/example/nextjs/lib/withI18next.js#L30

setting initialProps: `initialI18nStore` and `initialLanguage`

that is basically what make SSR work without rerender on client -> the store: https://github.com/i18next/react-i18next/blob/master/example/nextjs/i18n.js#L69

-> takes from the serverside i18next resource store the translation files/data and passes that down in initialProps as `initialI18nStore` -> the react-i18next withNamepaces or NamespacesConsumer take that props and fill the store instead of need to load -> insta ready to render without mismatch with server.

`initialLanguage` -> should be clear. https://github.com/i18next/react-i18next/blob/master/example/nextjs/i18n.js#L74

you might easily do the same in your app.js passing those two props into the NamespacesConsumer

**Comment by Enalmada** [comment]
I am happy to report that seems to have worked!

```
export default class MyApp extends App {

  static async getInitialProps({Component, router, ctx}) {
    ...
    let i18nProps = i18n.getInitialProps(ctx.req);
    return {pageProps, i18nProps};
  }

  render() {
    const { Component, pageProps, i18nProps } = this.props;

    return (
      <Container>
        <NamespacesConsumer
          {...pageProps}
          ns="common"
          i18n={(pageProps && pageProps.i18n) || i18n}
          wait={process.browser}
          initialI18nStore={i18nProps.initialI18nStore}
          initialLanguage={i18nProps.initialLanguage}
        >
...
```

Can I submit as a pull request (any changes you request to how I did it)?

**Comment by jamuhl** [comment]
sure PR's are welcome anytime...just do not change formatting...beside that ... all green ;)

**Comment by jamuhl** [comment]
hm just...that will only work in your case i guess

**Comment by jamuhl** [comment]
having more namespaces used in you need to pass namespaces as second argument to `i18n.getInitialProps(ctx.req);`

in your case it takes the default namespace (which is the only one)

**Comment by Enalmada** [comment]
Well I would like to make it work as best as possible for everyone.  Any hints as to what that second argument might look like / come from for the example code?

**Comment by jamuhl** [comment]
in the withI18next its solved with this: https://github.com/i18next/react-i18next/blob/master/example/nextjs/lib/withI18next.js#L20

guess same could be done here...the point is in the example provided here its not needed as all pages get wrapped withI18next

so this would be more a PR for a sample having no need for namespaces per page - so i'm not sure we should do an extra sample, or some sort of comments in the _app.js with hints how you solved it...idk

**Comment by Enalmada** [comment]
I totally agree that most pages will ultimately be wrapped with `withI18next`...but starting out trying to not wrapping them might be more common than you think.   I suspect most new users will start with a common namespace because that is how they have always done it (ie "en.properties") before amazing tools like react-i18next opened up the ability to have namespaces.  I happen to be migrating a site with a huge messages file and will break it up once I get "common" working like it was using the old pre-react technology.   I would bet I am not the first or the last new user to try passing NamespacesConsumer {t} down to existing pages rather than putting `withI18next("common")` on every page and fail and feel forced to add withI18next when perhaps it could have actually worked if NamespacesConsumer had something like the line you point out above.    

If you think you could easily solve the problem somehow in NamespacesConsumer and fix the currently broken page 3 ssr without any example code changes then that would provide the best user experience.  Otherwise I think for now I can checkin my fix with some notes why for users to avoid hitting the same issues I did.

**Comment by Enalmada** [comment]
@jamuhl Could you tell me what the relationship is with this example and the one in the next repo?  Should If I want to make an example change, do I do a pull request to both or are things migrated from here to there?

**Comment by jamuhl** [comment]
@Enalmada plan is to move the sample here back to next.js site as soon i got a good feeling this sample here works stable (the sample on next.js currently has some flaws - but it's a lot easier for me to work here) - even i'm not sure i should replace the sample there with a link to the sample here - would make it easier for closing issues / questions regarding next.js / react-i18next as i do not have a look on the next repo at all...
